File: src/main/java/de/pflugradts/passbird/application/configuration/ConfigurationFactory.java
package de.pflugradts.passbird.application.configuration;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.google.inject.Inject;
import de.pflugradts.passbird.application.util.SystemOperation;

import java.util.Optional;

import static de.pflugradts.passbird.application.configuration.ReadableConfiguration.CONFIGURATION_FILENAME;
import static de.pflugradts.passbird.application.configuration.ReadableConfiguration.CONFIGURATION_SYSTEM_PROPERTY;

public class ConfigurationFactory {

    @Inject
    private SystemOperation systemOperation;

    public UpdatableConfiguration loadConfiguration() {
        return Optional.ofNullable(configurationFromFile()).orElse(Configuration.createTemplate());
    }

    private UpdatableConfiguration configurationFromFile() {
        final var directory = System.getProperty(CONFIGURATION_SYSTEM_PROPERTY);
        final var mapper = new ObjectMapper(new YAMLFactory());
        try {
        return mapper.readValue(
                systemOperation.getPath(directory).resolve(CONFIGURATION_FILENAME).toFile(),
                Configuration.class);
        } catch (Exception ex) {
            // FIXME error handling
            return null;
        }
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/configuration/ConfigurationFactory.kt
package de.pflugradts.passbird.application.configuration
import de.pflugradts.kotlinextensions.tryCatching
import de.pflugradts.passbird.application.RunContext
import de.pflugradts.passbird.application.configuration.ReadableConfiguration.Companion.CONFIGURATION_FILENAME
import de.pflugradts.passbird.application.failure.ConfigurationFailure
import de.pflugradts.passbird.application.failure.reportFailure
import de.pflugradts.passbird.application.toFileName
import de.pflugradts.passbird.application.util.FAILURE_EXIT_STATUS
import de.pflugradts.passbird.application.util.SystemOperation
class ConfigurationFactory constructor(
    private val systemOperation: SystemOperation,
    private val runContext: RunContext,
    private val configurationYamlMapper: ConfigurationYamlMapper = ConfigurationYamlMapper(),
) {
    fun loadConfiguration() = if (!systemOperation.exists(filePath)) {
        Configuration(template = true)
    } else {
        configurationFromFile()
    }
    private fun configurationFromFile() = tryCatching {
        filePath.let {
            it.toFile().inputStream().use(configurationYamlMapper::readConfiguration)
        }
    }.let { result ->
        result.exceptionOrNull()?.let {
            reportFailure(ConfigurationFailure(it))
            systemOperation.exit(FAILURE_EXIT_STATUS)
            throw it
        }
        result.getOrNull()!!
    }
    private val filePath get() = systemOperation.resolvePath(
        runContext.homeDirectory,
        CONFIGURATION_FILENAME.toFileName(),
    )
}
